### PR TITLE
Operator New Alloc to external only

### LINKED ISSUE #350: Clip cloning leads to audible culling in simple songs
With very simple test song with one bassline of 16 steps (no detune) and a kit with 4 to the floor kick pattern and nothing else at 120 bpm I get audible culling and the following debug output:

```culled 1 voices. numSamples: 46. voices left: 1```

which can not be reproduced with official beta

### PR DESCRIPTION
**Pull request by m-m-adams**
Changes the new operator to allocate in external RAM. This frees a significant amount of internal RAM and regains some of our lost performance 

Note that with this change any performance critical objects must be manually allocated and initialized since new would place them on slower SDRAM. Luckily the audio related code already does this.

I believe this fixes  #350

### CODE REVIEW

File: src/deluge/memory/general_memory_allocator.cpp [hunk 1]
```diff
@@ -140,7 +141,15 @@ uint32_t GeneralMemoryAllocator::getAllocatedSize(void* address) {
 }
 
 int32_t GeneralMemoryAllocator::getRegion(void* address) {
-	return ((uint32_t)address >= (uint32_t)INTERNAL_MEMORY_BEGIN) ? MEMORY_REGION_INTERNAL : MEMORY_REGION_SDRAM;
+	if ((uint32_t)address >= (uint32_t)INTERNAL_MEMORY_BEGIN) {
+		return MEMORY_REGION_INTERNAL;
+	}
+	else if (((uint32_t)address <= EXTERNAL_MEMORY_BEGIN) || (uint32_t)address >= EXTERNAL_MEMORY_END) {
+		numericDriver.freezeWithError("e999");
```
Comment (bobtwinkles) [inline src/deluge/memory/general_memory_allocator.cpp:148]: I think we should start a convention of using "Mxxx" as the error codes in the allocator (the smoke test in #351  uses M000 for example)

File: src/deluge/memory/general_memory_allocator.cpp [hunk 2]
```diff
@@ -140,7 +141,15 @@ uint32_t GeneralMemoryAllocator::getAllocatedSize(void* address) {
...
+		return -1;
+	}
+	return MEMORY_REGION_SDRAM;
+	//return ((uint32_t)address >= (uint32_t)INTERNAL_MEMORY_BEGIN) ? MEMORY_REGION_INTERNAL : MEMORY_REGION_SDRAM;
```
Comment (bobtwinkles) [inline src/deluge/memory/general_memory_allocator.cpp:152]: can delete this I think

File: src/deluge/memory/memory_region.h [hunk 3]
```diff
@@ -38,7 +38,7 @@ struct NeighbouringMemoryGrabAttemptResult {
 #define SPACE_HEADER_ALLOCATED 0x80000000
 
 #define SPACE_TYPE_MASK 0xC0000000
-#define SPACE_SIZE_MASK 0x3FFFFFFF
+#define SPACE_SIZE_MASK 0x3FFFFFFFu
```
Comment (bobtwinkles) [inline src/deluge/memory/memory_region.h:41]: Should add the `u` suffix to TYPE_MASK as well, as long as we're in the area

File: src/deluge/memory/memory_region.h [hunk 4]
```diff
@@ -54,7 +54,7 @@ class MemoryRegion {
 	void dealloc(void* address);
 	void verifyMemoryNotFree(void* address, uint32_t spaceSize);
 
-	OrderedResizeableArrayWithMultiWordKey emptySpaces;
+	//OrderedResizeableArrayWithMultiWordKey emptySpaces;
```
Comment (bobtwinkles) [inline src/deluge/memory/memory_region.h:57]: delete?

File: src/resetprg.c [hunk 5]
```diff
@@ -144,6 +144,15 @@ void resetprg(void) {
...
 	__enable_irq();
 	__enable_fiq();
+	// Setup SDRAM. Have to do this before we init global objects
+	userdef_bsc_cs2_init(0); // 64MB, hardcoded
+#if !defined(NDEBUG)
+	const uint32_t SDRAM_SIZE = EXTERNAL_MEMORY_END - EXTERNAL_MEMORY_BEGIN;
+	uint16_t* sdram = (uint16_t*)EXTERNAL_MEMORY_BEGIN;
+	for (size_t i = 0; i < SDRAM_SIZE; i++) {
+		sdram[i] = 0x0000;
+	}
```
Comment (bobtwinkles) [inline src/resetprg.c:154]: using SIMD or a `memset` might be faster here? Or is there a particular reason to use `uint16` sized writes

File: src/deluge/memory/memory_region.h (hunk 3 above)
Comment (m-m-adams) [reply src/deluge/memory/memory_region.h:41]: Yeah sure, I left it since that one defaults to unsigned already

File: src/resetprg.c (hunk 5 above)
Comment (m-m-adams) [reply src/resetprg.c:154]: Memset is faster, I had it in one of earlier commits but I guess it got lost along the way. That explains why it got slower

**Comment by PaulFreund** [comment]
I tested this PR in RelWithDebInfo and audibly it is the same unfortunately but in RTT instead of hard culling I get
```soft-culled 1 voice. voices now: 2```
scrolling through loading presets still gets me some hard culling. I think it is a step in the right direction and in combination with reducing image size could be enough. Once we get the same performance as our baseline we probably need to feed the audio rendering in those functions.

File: src/deluge/memory/memory_region.cpp [hunk 6]
```diff
@@ -98,7 +99,10 @@ static EmptySpaceRecord* recordToMergeWith;
 // Specify the address and size of the actual memory region not including its headers, which this function will write and don't have to contain valid data yet.
 // spaceSize can even be 0 or less if you know it's going to get merged.
 inline void MemoryRegion::markSpaceAsEmpty(uint32_t address, uint32_t spaceSize, bool mayLookLeft, bool mayLookRight) {
-
+	if ((address <= start) || address >= end) {
+		numericDriver.freezeWithError("e999");
```
Comment (bobtwinkles) [inline src/deluge/memory/memory_region.cpp:103]: -> `M998`

**Review by bobtwinkles — APPROVED** [review]
minor nitpick but otherwise LGTM

**Pull request merged**